text/0000-local-loop-bindings.md

```diff
@@ -0,0 +1,218 @@
+- Feature Name: local_loop_bindings
+- Start Date: 2018-12-25
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This will allow an extended syntax for `loop` to accept local variables, which change for each iteration to simplify more complicated loop constructs and avoid mutable state.
+
+
+# Motivation
+[motivation]: #motivation
+
+The new syntax is inspired by `loop` in the upcoming release of the [scopes programming language](scopes.rocks).
+This way it's possible use different values for each iteration without the need of mutable variables defined outside of the loop.
+
+The variables will be defined after the loop keyword, so they will only be accessible in the scope of the loop, not afterwards. They will not be mutable by default, so it can be ensured, that the variables only change once per iteration.
+
+Especially since loops can return values, it's not necessary at all to mutate state inside a loop in some cases.
+
+This is a more functional programming style, which may also allow more optimizations like storing the loop arguments in registers instead of allocating storage for mutable variables.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The extended syntax for `loop` will just work like a `while let`. "while let" will be replaced by "loop". Unlike `while let`, the pattern is not allowed to be refutable, so enum variants are not allowed on the lefthand side of "=".
+The introduced bindings will be accessible in this loop.
+
+The new syntax will look like this:
+
+```rust
+loop binding = value {
+    /* body */
+}
+```
+
+The return value of the loop body will implicitely be passed to the next iteration of the loop, so it needs to be the same type as the initial value.
+
+An example of a simple iteration, which iterates the loop ten times and prints the iteration number would look like this:
+
+```rust
+loop i = 1 {
+    if i <= 10 {
+        println!("iteration {}", i);
+        i + 1
+    } else {
+        break;
+    }
+}
+```
+
+`continue` will accept an argument in this loop, which will be passed to the next iteration. Using continue, this could look like this:
+
+```rust
+loop i = 1 {
+    if i <= 10 {
+        println!("iteration {}", i);
+        continue i + 1;
+    }
+    break;
+}
+```
+
+Since the end of the loop is never reached, the return value is not required to be the type of the binding, here.
+
+A loop without bindings (`loop { /* body */ }`) will be the same as this:
+
+```rust
+loop () = () {
+    /* body */
+}
+```
+
+This will not be a breaking change, since it's not allowed to have values other than `()` from a loop.
+
+A simple example from the book looks like this:
+
+```rust
+let mut x = 5;
+let mut done = false;
+
+while !done {
+    x += x - 3;
+
+    println!("{}", x);
+
+    if x % 5 == 0 {
+        done = true;
+    }
+}
+```
+
+Using the new syntax, this could be rewritten as this:
```

Comment:
Or rewritten as the simpler:
```rust

loop (mut x, done) = (5, false) {
    if done {
        break;
    }
    x += x - 3;
    println!("{}", x);
    (x, x % 5 == 0)
}
```

Reply:
oh, why didn't I get this idea ^^
I thought, I'd have to write (x, if x % 5 == 0 {true} else {false}), and didn't like it...